erts/emulator/beam/erl_bif_lists.c

```diff
@@ -56,6 +56,39 @@ static int append_ctx_bin_dtor(Binary *context_bin) {
     return 1;
 }
 
+/* Scans a list to check whether it is improper or proper, yielding.
+ * Returns 1 if the list is proper, -1 if it is improper or 0 when yielding. */
+static int append_is_proper_list(Process *p, Eterm *iterator_p) {
+    static const Sint ELEMENTS_PER_RED = 32;
+    Sint budget, count;
+    Eterm iterator;
+
+    budget = ELEMENTS_PER_RED * ERTS_BIF_REDS_LEFT(p);
+    iterator = *iterator_p;
+
+#ifdef DEBUG
+    budget = budget / 10 + 1;
+#endif
+
+    for (count = 0; count < budget && is_list(iterator); count++) {
+        iterator = CDR(list_val(iterator));
+    }
+
+    if (!is_list(iterator) && !is_nil(iterator)) {
+        return -1;
+    }
+
+    BUMP_REDS(p, count / ELEMENTS_PER_RED);
+
+    *iterator_p = iterator;
+
+    if (is_nil(iterator)) {
+        return 1;
+    }
```

Comment:
We should still bump reductions on errors. How about?

```suggestion
    BUMP_REDS(p, count / ELEMENTS_PER_RED);

    if (is_list(iterator)) {
        *iterator_p = iterator;
        return 0;
    } else if (is_nil(iterator)) {
        return 1;
    }

    return -1;
```